### PR TITLE
feat(builder): Add Supabase Auth, session and sign in UI

### PR DESCRIPTION
**Pull request by kcze**
This PR adds Supabase Auth (cloud) integration, login and profile UI, configures password login and three OAuth providers: Google, GitHub and Discord.

For `Account` button to show up and ability to login two env vars need to be set in `.env.local`: `NEXT_PUBLIC_SUPABASE_URL` and `NEXT_PUBLIC_SUPABASE_ANON_KEY`. OAuth providers are by the Supabase and don't require env vars.

Email confirmation (for email/password signup) is disabled because there's limit of 3 emails per hour without custom SMTP server configuration. [Link](https://supabase.com/dashboard/project/adfjtextkuilwuhzdjpf/auth/templates) to configure custom SMTP server and email template.

### Added dependencies:
- "@supabase/ssr": "^0.4.0"
- "@supabase/supabase-js": "^2.45.0"
- "react-icons": "^5.2.1"

### Added pages/routes:
- `app/auth/auth-code-error/page.tsx`: displayed when login using OAuth provider fails
- `app/auth/callback/route.ts`: route accessed when logging in using OAuth provider; it passes session code to Supabase
- `app/auth/confirm/route.ts`: accessed when confirming email, users will be directed here from email they get after signing in.
- `app/error/page.tsx`: Generic error page without explanation (any errors should be visible in the console)
- `app/login/page.tsx` and `app/login/actions.ts`: Login page and related login/signup server actions
- `app/profile/page.tsx`: Profile page, displays email address of the user and button to logout

### Changes
- Update `layout.tsx`: add `Log In` button and make icons consistent. The log in button shows up if user is logged out, avatar is shown when logged in, and if supabase is unavailable nothing shows up.
- Login form is verified using `zod` on the frontend (recommended by shadcn) and in case login fails feedback is displayed. On successful login users are redirected to `/profile`
- Add `PasswordInput` component, [source](https://gist.github.com/mjbalcueva/b21f39a8787e558d4c536bf68e267398)
- Add `SupabaseProvider` with context for Supabase accessed via hook `useSupabase(): { supabase: SupabaseClient | null, isLoading: boolean }`
- Add `useUser` hook to get `{ user, session, isLoading, error }` on the client 
- Add `getServerUser`: async function to get `{ user: User | null, error: string | null }` on the server side
- Add `src/middleware.ts` and `client.ts`, `server.ts`, `middleware.ts` in `src/lib/supabase` which are utility functions and middleware to refresh auth token

### CODE REVIEW

File: rnd/autogpt_builder/.env.example [hunk 1]
```diff
@@ -1 +1,7 @@
 AGPT_SERVER_URL=http://localhost:8000/api
+
+## Supabase credentials
+## YOU ONLY NEED THEM IF YOU WANT TO USE SUPABASE USER AUTHENTICATION
+## If you're using self-hosted version then you most likely don't need to set this
+# NEXT_PUBLIC_SUPABASE_URL=your-project-url
+# NEXT_PUBLIC_SUPABASE_ANON_KEY=your-anon-key
```
Comment (ntindle) [inline rnd/autogpt_builder/.env.example:7]: is it possible to default these to our instance safely

File: rnd/autogpt_builder/src/app/login/page.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  if (!supabase) {
+    return <div>User accounts are disabled because Supabase client is unavailable</div>
+  }
+
+  async function handleSignInWithProvider(provider: 'google' | 'github' | 'discord') {
+    const { data, error } = await supabase!.auth.signInWithOAuth({
+      provider: provider,
+      options: {
+        redirectTo: `http://localhost:3000/auth/callback`,
```
Comment (aarushik93) [inline rnd/autogpt_builder/src/app/login/page.tsx]: is this right? can this be in an env var?

**Comment by aarushik93** [comment]
btw there's a logout option in the header where the profile photo is - maybe we can update that?
Screenshot 2024-08-02 at 10 47 07 AM

File: rnd/autogpt_builder/.env.example (hunk 1 above)
Comment (kcze) [reply rnd/autogpt_builder/.env.example:7]: I don't think so

File: rnd/autogpt_builder/src/app/login/page.tsx (hunk 2 above)
Comment (kcze) [reply rnd/autogpt_builder/src/app/login/page.tsx]: I made it an env var with `http://localhost:3000/auth/callback` as default
edit: I tried `/auth/callback` but it doesn't work so there must be a full url

**Comment by kcze** [comment]
> btw there's a logout option in the header where the profile photo is - maybe we can update that?

Updated, now if user is logged out there's `Log In` button and when logged in an avatar with working buttons.

**Pull request merged**